fix: support multiple systems

ml1nk opened the pull request:
Improve names (include home.name) and unique_ids (include system.id) to support multiple accounts or single accounts with multiple systems on it.

In addition:
- zone names include a reference to the matching circulation if exists (couldn't match them before)
- only calculate energy efficiency if self.energy_consumed isn't None and greater than 0 (it crashed after midnight when I was developing)

After merging you should remove the integration and add it again, as every id is changed.

Fixes #15

ml1nk commented:
@signalkraft I'm currently changing some more code to support multiple headpumps in one system, which separates all devices. Do you want a separate pull request or an update to this one?

signalkraft commented on custom_components/mypyllant/climate.py:
```diff
@@ -269,21 +270,28 @@ def system(self) -> System:
         return self.system.zones[self.zone_index]
 
+    @property
+    def circuitNameSuffix(self) -> str:
```
Please use snake_case instead of lowerCamelCase

signalkraft commented on custom_components/mypyllant/sensor.py:
```diff
@@ -292,9 +296,20 @@ def system(self) -> System:
         return self.system.zones[self.zone_index]
 
+    @property
+    def circuitNameSuffix(self) -> str:
```
Please use snake_case instead of lowerCamelCase

signalkraft commented on custom_components/mypyllant/utils.py:
```diff
@@ -12,10 +13,15 @@ def get_system_sensor_unique_id(system: System) -> str:
+def get_unique_id_prefix(system_id: str) -> str:
+    return f"{DOMAIN}_{system_id}_"
+
+def get_name_prefix(home: str) -> str:
```
I think you still need to install pre-commit and run it on your changes. I've merged your other PR and updated the pre-commit settings and VS Code lint task.

signalkraft reviewed (commented):
Thank you so much for the contribution, I added a few comments!

I think it would make sense to add the changes for multiple heatpumps in the same PR. It'll get pretty big, but it's all related.

If you can, please also add some test data to the underlying mypyllant library, that would help ensure that your changes don't break with future update.

signalkraft commented:
Also it would be great, if you can figure out how to [migrate the entity IDs](https://community.home-assistant.io/t/migrating-unique-ids/348512) to the new format. That way existing dashboards and automations (probably?) wouldn't break for users.

Alternatively, you could use the new format only for users with multiple homes / heatpumps. But that would make the code for unique IDs more complicated.

ml1nk commented:
> Also it would be great, if you can figure out how to [migrate the entity IDs](https://community.home-assistant.io/t/migrating-unique-ids/348512) to the new format. That way existing dashboards and automations (probably?) wouldn't break for users.
> 
> Alternatively, you could use the new format only for users with multiple homes / heatpumps. But that would make the code for unique IDs more complicated.

I think migrating the existing entities/devices is the only feasible way. I will try to write something up, after anything else is finished.

ml1nk commented:
New Changes:
- Every device is listed separately out of two reasons:
   - without test data with two heat pumps i don't know how to match the hydraulik station to the correspondant head pump
   - i found it interesting to know when the heater is running and how much power it needs
- EfficiencySensor is created for every device and one for the system itself 
- i fixed or commented out some type errors, without moving more code and enabled basic type checking

ToDos:
- get vscode pytest integration to run correctly inside the dev container  (currently socket.socket error)
- get the linter to work inside vscode and by command outside of git commits (thanks for your help) and fix all of the remaining warnings

Future Ideas:
- show the primary heat source as attribute / reference (don't know what ha provides here)
- show the error codes also in the devices they are occurring instead of a list in the system device 

What is your opinion on the naming? I changed the naming of all entities and sensors to be unique but this leads to rather long names. On the opposite it is hard to find the correct elements if they are short. Currently I prefer shorten all entities but let the devices unique, but I'm unsure about it.

Fixes #52

signalkraft commented:
> What is your opinion on the naming? I changed the naming of all entities and sensors to be unique but this leads to rather long names.

I personally also prefer shorter IDs, to be able to see the full entity in the Home Assistant UI. You could store a list of known system IDs in `self.hass.data[DOMAIN][self.entry.entry_id]["system_ids"]` and use the position of the system ID in the list instead of the value. Whenever there's a new system or home added, the new value gets appened. To make it more human-readable, you could add a shortened system name or home name to it.

signalkraft commented:
> You mean by Contributing Test Data like explained here https://github.com/signalkraft/myPyllant? I tried it before (https://github.com/signalkraft/mypyllant-component/issues/15#issuecomment-1848795279) but it doesn't worked for me. I can try again if that's what you meant.

Yes, please try again. If the error occurs again, I'll have another look at the log and see if I can fix it.

ml1nk commented:
> > What is your opinion on the naming? I changed the naming of all entities and sensors to be unique but this leads to rather long names.
> 
> I personally also prefer shorter IDs, to be able to see the full entity in the Home Assistant UI. You could store a list of known system IDs in `self.hass.data[DOMAIN][self.entry.entry_id]["system_ids"]` and use the position of the system ID in the list instead of the value. Whenever there's a new system or home added, the new value gets appened. To make it more human-readable, you could add a shortened system name or home name to it.

Just to clarify: We are talking about the name shown in the interface of the entity/device not the unique_id. The name is used as a base for the entity_id (which doesn't need to be set manually and has a counter if a duplicate occurs). The unique_id is never shown, at least I couldn't find it in the UI. 

In case of @andreazame the position of the system_ids isn't unique (two entries means two indexes) and to be unique probably needed to be combined with the self.entry.entry_id. The system_id isn't used in the name currently, instead home_name is always prefixed.

Currently the name is set to the following.

Devices:
- System/Gateway: {home_name}
- Devices: {home_name} {self.de_index} {self.device.name_display}
  - de_index => device index (differentiate multiple heatpumps of the same type)
- Zones: {home_name} Zone {self.zone.name} of Circuit {self.zone.associated_circuit_index}
  -  "of Circuit" only  if associated
- Circuits: {home_name} Circuit {self.circuit_index}
- Ventilation: {home_name} Ventilation {vname}
- Domestic Hot Water: {home_name} Domestic Hot Water {self.dhw_index}

Entities:
- {home_name} Trouble Codes on {self.system.system_name}
- {home_name} Online Status
- {home_name} Firmware Update Required
- {home_name} Firmware Update Enabled
- {home_name} Cooling Allowed in {self.circuit_index}
- {home_name} Climate Zone {self.zone.name}
- {home_name} Ventilation {vname}
- {home_name} Outdoor Temperature
- {home_name} System Water Pressure
- {home_name} Firmware Version
- {home_name} Desired Temperature in {self.zone.name}
- {home_name} Current Temperature in {self.zone.name}
- {home_name} Humidity in {self.zone.name}
- {home_name} Heating Operating Mode in {self.zone.name}
- {home_name} Heating State in {self.zone.name}
- {home_name} Current Special Function in {self.zone.name}
- {home_name} Current Flow Temperature in Circuit {self.circuit.index}
- {home_name} State in Circuit {self.circuit.index}
- {home_name} Min Flow Temperature Setpoint in Circuit {self.circuit.index}
- {home_name} Heating Curve in Circuit {self.circuit.index}
- {home_name} Tank Temperature Domestic Hot Water {self.domestic_hot_water.index}
- {home_name} Setpoint Domestic Hot Water {self.domestic_hot_water.index}
- {home_name} Operation Mode Domestic Hot Water {self.domestic_hot_water.index}
- {home_name} Current Special Function Domestic Hot Water {self.domestic_hot_water.index}
- {home_name} {self.de_index} {name} {et}{om}
  - om=> operation_mode (if exists)
  - et => energy_type
  - de_index => device index
- {home_name} {self.de_index} {self.device_data_list[0].device.name_display} Heating Energy Efficiency
  -  de_index => device index 
  - {home_name} Heating Energy Efficiency if system wide
- {home_name} Water Pressure {self.device.name}
- {home_name} Domestic Hot Water {self.dhw_index}

As you normally choose devices and can filter entities by room i would prefer to remove all information out of the entity name which is already in the correspondent device. So no home_name, zone_name, device_name or index anymore. What do you think?

signalkraft commented:
> Just to clarify: We are talking about the name shown in the interface of the entity/device not the unique_id.

My bad! I think suggestion for names makes sense. As for unique IDs, I just tested how that works in your PR and understood it now. Also seems fine!

> Did you know that behavior and would it also be feasible in your optinion?

Did not know, and seems like a good solution!

ml1nk commented:
> > Just to clarify: We are talking about the name shown in the interface of the entity/device not the unique_id.
> 
> My bad! I think suggestion for names makes sense. As for unique IDs, I just tested how that works in your PR and understood it now. Also seems fine!
> 
> > Did you know that behavior and would it also be feasible in your optinion?
> 
> Did not know, and seems like a good solution!

Lucky me, I just finished a rework for names and unique_ids. Outside of using prefixes for entities and reworked the unique_id to be matching, i changed the usage of system.home.name to system.home.home_name (if it exists) as it is shorter.

If you are happy with the results, I would start trying to write a migration. Can you update the beta, please?

ml1nk commented:
BTW: @signalkraft Are you at the 37c3?

signalkraft commented:
I was on the old one, I've created a new pre-release with the latest commits and will check it out.

On first glance the names seem off in a few places for me:

![Screenshot from 2023-12-29 19-27-37](https://github.com/signalkraft/mypyllant-component/assets/3762829/4ad2b868-4460-45c7-bf69-3fc496e0c56e)

* In some entities the "home" doesn't get removed
* `Home Zone Zone 1 of Circuit 0 Climate` seems a little nightmarish, maybe `Zone 1 (Circuit 0) in Home`? I doesn't need to be perfect, users can rename entities, but at least a little more readable would be neat
* `Device <Number>` seems unnecessary to me in the names, but maybe I missed why it's necessary?

ml1nk commented:
> I was on the old one, I've created a new pre-release with the latest commits.
> 
> On first glance the names seem off in a few places for me:
> 
> ![Screenshot from 2023-12-29 19-27-37](https://private-user-images.githubusercontent.com/3762829/293408925-4ad2b868-4460-45c7-bf69-3fc496e0c56e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDM4NzUxNzYsIm5iZiI6MTcwMzg3NDg3NiwicGF0aCI6Ii8zNzYyODI5LzI5MzQwODkyNS00YWQyYjg2OC00NDYwLTQ1YzctYmY2OS0zZmM0OTZlMGM1NmUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDIzMTIyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMjlUMTgzNDM2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjY2YWEyZjZkZThlZWZkZjBiOTM2Mzc1MDU3ZDRiZGJlOTBkZjNiYzU3NzQ5Yzk3YzhmNGNhYTExMWZmOTQ4NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.lQEUsupjEWKcW_4E0-5bRGSx-3KUHndfkS5_s1DC3fs)
> 
> * In some entities the "home" doesn't get removed
> * `Home Zone Zone 1 of Circuit 0 Climate` seems a little nightmarish, maybe `Zone 1 (Circuit 0) in Home`? I doesn't need to be perfect, users can rename entities, but at least a little more readable would be neat
> * `Device <Number>` seems unnecessary to me in the names, but maybe I missed why it's necessary?

The first two points are my mistake, i will fix it. I added Device <Number> to differentiate between different Devices of the same type on a system, like it is expected in #52.

ml1nk commented:
Changes:
- DOMAIN removed from Outdoor Temperature
- of Circuit <number> => (Circuit <number>)
  - still the home.home_name (Home in your case, the street name in mine) always comes first, so that the device name is a prefix of the entity name
- Rename entity "Home Domestic Hot Water 0" to  "Home Domestic Hot Water 0 Climate", so that it is shown as Climate

My Zones are named like "Fbh" or "Hk", so in my case the prefix Zone makes sense, but if you want i can remove it either way.

signalkraft commented:
> My Zones are named like "Fbh" or "Hk", so in my case the prefix Zone makes sense, but if you want i can remove it either way.

Maybe check if the name starts with `Zone` and leave the prefix out in that case? My guess is that quite a few people use the standard names.

ml1nk commented:
> > My Zones are named like "Fbh" or "Hk", so in my case the prefix Zone makes sense, but if you want i can remove it either way.
> 
> Maybe check if the name starts with `Zone` and leave the prefix out in that case? My guess is that quite a few people use the standard names.

Good idea, does it work?

ml1nk commented:
> > You mean by Contributing Test Data like explained here https://github.com/signalkraft/myPyllant? I tried it before ([#15 (comment)](https://github.com/signalkraft/mypyllant-component/issues/15#issuecomment-1848795279)) but it doesn't worked for me. I can try again if that's what you meant.
> 
> Yes, please try again. If the error occurs again, I'll have another look at the log and see if I can fix it.

I'm sorry, but i got the same error as before.

signalkraft commented:
> I'm sorry, but i got the same error as before.

Could you retry with the latest version I pushed to main at https://github.com/signalkraft/myPyllant ?

signalkraft commented:
> > Maybe check if the name starts with `Zone` and leave the prefix out in that case? My guess is that quite a few people use the standard names.
> 
> Good idea, does it work?

Jup, perfect!

ml1nk commented:
> > I'm sorry, but i got the same error as before.
> 
> Could you retry with the latest version I pushed to main at https://github.com/signalkraft/myPyllant ?

Worked perfectly, here is my testdata:
[twosystems.tar.gz](https://github.com/signalkraft/mypyllant-component/files/13796510/twosystems.tar.gz)

ml1nk commented:
I got the basic migration method working, but it is very tricky to do the job correctly for all entities and devices. In addition some entities are now matched to other devices and there even the auxiliary heater is now separated. If you are ok with it, i would prefer make it a breaking chance and commit only the preparations to do migrations in the future for easier cases.

Are you happy with all identifiers and unique_ids

signalkraft commented:
Sure, sounds reasonable. 

> Are you happy with all identifiers and unique_ids

Yes, they look good to me.

signalkraft commented:
I'm finishing some test cases with our test data and waiting for feedback in the ticket. When you think this is ready to merge and tested well enough by the people in #52 let me know.

ml1nk commented:
Perfect. I will update my own installation and test if it all works like before and hopefully we will get #52 solved today.

BTW: I changed export.data selector to boolean as it failed the hassfest validator before.

signalkraft commented:
I've merged your PR and created a new release with the changes to the myPyllant library as well.
